src/openfermion/ops/_quadratic_hamiltonian.py

```diff
@@ -0,0 +1,104 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""Class and functions to store and manipulate Hamiltonians that are quadratic
+in the fermionic ladder operators."""
+from __future__ import absolute_import
+
+import numpy
+
+from openfermion.config import EQ_TOLERANCE
+from openfermion.ops import PolynomialTensor
+
+
+class QuadraticHamiltonianError(Exception):
+    pass
+
+
+class QuadraticHamiltonian(PolynomialTensor):
+    """Class for storing Hamiltonians that are quadratic in the fermionic
+    ladder operators. The operators stored in this class take the form::
+
+        \sum_{p, q} (M_{pq} - \mu \delta_{pq}) a^\dagger_p a_q
+        + 1 / 2 \sum_{p, q} (A_{pq} a^\dagger_p a^\dagger_q + h.c.)
+        + constant
+
+    where
+        M is a Hermitian n_qubits x n_qubits matrix.
+        A is an antisymmetric n_qubits x n_qubits matrix.
+        \mu is a float representing the chemical potential term.
+        \delta_{pq} is the Kronecker delta symbol.
+
+    We separate the chemical potential \mu from M so that we can use it
+    to adjust the expectation value of the total number of particles.
+
+    Attributes:
+        n_qubits: An int giving the number of qubits.
+        constant: A constant term in the operator given as a float.
+    """
+
+    def __init__(self, constant, hermitian_part,
+                 antisymmetric_part=None, chemical_potential=None):
+        """
+        Initialize the QuadraticHamiltonian class.
+
+        Args:
+            constant(float): A constant term in the operator.
```

Comment:
So, I am wondering if the constant should be stored (in PolynomialTensor) as the entry in the dictionary that is represented by (). This makes sense to me since that is how the constant is stored for FermionOperator. What do you think?

Reply:
I thought about that too, and the only reason I didn't implement that is that then the line
`self.n_qubits = list(n_body_tensors.values())[0].shape[0]`
wouldn't work, since `list(n_body_tensors.values())[0]` might return the constant, which is a float. However, we could also just remove the `()` entry of `n_body_tensors` when __init__ is called, and replace it with the constant passed in. That just seemed kind of ugly to me.

Reply:
Actually, I think I do have a better solution; I can do it later or submit in another PR.